### ISSUE TITLE
autocalc rec len for MS3Record

### ISSUE THREAD

**Issue by crotwell** [EarthScope/mseedlib#3]
This fails:

```
#!/usr/bin/env python3

from mseedlib import MS3Record
rec = MS3Record()
rec.formatversion = 3
rec.encoding = 3   # int32
rec.set_starttime_str("2024-01-01T15:13:55.123456789Z")
rec.samplerate = 40.0
#rec.reclen = 999999
data = [(i%99-49) for i in range(0,1000)]

def record_handler(buffer, handlerdata):
    print(f"Got {len(buffer)} bytes")

rec.pack(record_handler, handlerdata={}, datasamples=data, sampletype='i', verbose=1)
```
with
```
Error: : Record length is out of range: 0
Traceback (most recent call last):
  File "/Users/crotwell/Code/seis/ms3eh/python/simpleCreateData.py", line 15, in <module>
    rec.pack(record_handler, handlerdata={}, datasamples=data, sampletype='i', verbose=1)
  File "/opt/homebrew/Caskroom/miniforge/base/envs/ms3eh/lib/python3.11/site-packages/mseedlib/msrecord.py", line 419, in pack
    raise MseedLibError(packed_records, f'Error packing miniSEED record(s)')
mseedlib.exceptions.MseedLibError: Generic error :: Error packing miniSEED record(s)
```
but adding:
```
rec.reclen = 999999
```
works, but seems unnecessary.

For miniseed2, a record length makes sense, and probably also for MSTraceList. But for a single MS3Record, there is no problem with arbitrary large records, so would be nice if MS3Record.pack() should just calculate the record size itself when reclen is the default 0. If I ask for a single record and it is really large due to a really long data array, then that is my own fault.

**Comment by chad-earthscope** [comment]
> But for a single MS3Record, there is no problem with arbitrary large records
Except that the library/mseedlib will not be able to read them. 

But, yeah, the default record length should have been used in that example.

**Comment by crotwell** [comment]
Is there a max record length for the library to read? This is only 1000 points, so seems like it isn't crazy big.

I think in many instances a single large record makes more sense that a whole bunch of pieces, at least when you already have a big array of data in memory. Be nice if the library could deal with that case.

**Comment by chad-earthscope** [comment]
> Is there a max record length for the library to read? This is only 1000 points, so seems like it isn't crazy big.

Yes, for the current release, [defined in libmseed.h](from libmseed.h):
```C
#define MAXRECLEN 10485760 //!< Maximum miniSEED record length supported (10MiB)
#define MAXRECLENv2 131172 //!< Maximum v2 miniSEED record length supported (131+ KiB or 2^17)
```

This probably isn't the source of the issue, but is a counter-point to the assertion that "... for a single MS3Record, there is no problem with arbitrary large records".

limbseed has the concept of a default record length used when not set by the caller.  Regardless of the default or user-specified, the record length is fixed for miniSEED v2 and becomes the maximum for v3.  One of the issues for mseedlib appears to be that the default is not working.

**Comment by crotwell** [comment]
True, let me rephrase:
> But for a single MS3Record, the spec allows for very large records

Use case:

I have an array of integers in memory that represents a large but within spec contiguous timeseries. I would like to save it as a single miniseed3 record without having to guess at the final saved byte size. I accept responsibility if the resulting file is ginormous. :)

**Comment by chad-earthscope** [comment]
Context for future readers: the discussion of large record support is not specific to mseedlib (this repository) but to libmseed's support of the specification.

> But for a single MS3Record, the spec allows for very large records

libmseed, and therefore mseedlib, do not support the full range of the specification for this reason: the design of software meant to handle relatively small records can be simplified and optimized significantly compared to a design intended to handle record lengths of ~4 GiB.  For example, assuming that a record can be read wholly into memory without potentially impacting the system makes for a much simpler & optimized design.  To that point libmseed re-uses a single pre-allocated buffer for reading records, which is enormously beneficial in that it reduces allocations and allows the decoding logic to operate on a single buffer.  Without this, the reading and decoding logic would require stateful "paging" through partial views of a record incrementally fetched from storage/network/etc, which is a non-trivial complexity.  I would argue that this kind of "paging" through a ginormous record has very little benefit over reading smaller records with the same data. 

From a practical perspective, I do not think any real use is seriously hampered by partitioning a continuous time series into multiple records.  Anyone using a miniSEED library (as they should), would not care or even know that the time series was partitioned/chunked/split into records.  When partitioning gigabytes of data into, say, 1 MiB records, the amount of overhead in volume and data series reconstruction is pretty small to the point of non-relevance.

> I would like to save it as a single miniseed3 record without having to guess at the final saved byte size.

My argument above is that the vast majority of users do not really care how many records it is.  Regarding guessing the final byte size, I'm guessing that if the default maximum record size of 4096 bytes had worked, most people would not even notice that more than one record was produced.  If you would like to be pedantic about record size you're welcome to set the length up to 10 MiB and create large (but not ginormous) records.

The miniSEED v2 default size of 4096 bytes makes a lot of sense for broad compatibility.  An argument could be made for having a larger default maximum size for v3 records, like 1 MiB for example, or even upping the  libmseed maximum of 10 MiB.  But not 4 GiB as that would require a very different design, and even more complexity to avoid that design effecting the performance of the majority, small-record, use case.

**Comment by rcasey-earthscope** [comment]
Hi Folks-

	I am also reminded of clock drift for sampling, where record breaks allow for milepost assertions of current time before epsilon gets too high.  I could see where kilohertz sample rates might beg for longer records since you can record so much within reasonable time tolerances.  Still, is it really helpful to have such a very long record (as Chad indicates)?  Getting into the megabyte scale of a sample array will challenge most code and libraries as it represents a significant buffer size.  If the data is compressed, the application will likely need to internalize the entire record content just to pull out a decompressed range (unless the compression format is partitioned).

	Is there a use case we are looking at where really long arrays make sense?

	Just curious.

	-Rob

**Comment by crotwell** [comment]
**TL;DR**  For those citizens of the future that still wish to create large records, even if the above would suggest it is a _bad idea_ (TM) please have a look at the "competition", [simplemseed](https://pypi.org/project/simplemseed/). Pure python, no limits, plenty of rope to shoot yourself in the foot, etc.

Here is an example [of writing a ginormous record](https://github.com/crotwell/simplemseed/blob/main/examples/ginormous_ms3.py).

IMHO, there is a lot of grey area in between my slightly tongue-in-check ginormous array and 4096. Yes, if the array is so large it has trouble fitting into memory, then that is a problem. But it is not really that uncommon for say ambient noise folks to deal in 24 hour segments of data. And while that is large, it is not so large as to be trouble fitting into memory.

My point largely is that in the post-processing, science side of data storage, it is a lot simpler conceptually and practically to just store one contiguous float array into a single record similar to the old SAC file format, especially if you have just processed it as a single float array in memory. And you get the advantage of a single extra header json instead of having to decide to put stuff in every record or only in the first, or how to merge multiple extra header copies on read, etc. People get lots of data in miniseed2, but as soon as they do any processing, I think they tend to save as something else that can hold a big ol array.

Just for giggles, one day at 40sps is 86400*40=3,456,000 samples and comes out to about 13Mb. 

Just my $0.02.
